# Introduction to Deep Learning and Neural Networks with Keras

## Peer graded assignment

In [1]:
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

## Explore the dataset to understand it better

In [3]:
concrete_data.shape

(1030, 9)

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

## Get the target variable and the predictors of the data

In [5]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [6]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [7]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

## Normalise the predictors

In [8]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [9]:
X  = predictors_norm.values

In [10]:
X

array([[ 2.4767117 , -0.85647182, -0.8467326 , ...,  0.86273513,
        -1.2170788 , -0.27959729],
       [ 2.4767117 , -0.85647182, -0.8467326 , ...,  1.05565137,
        -1.2170788 , -0.27959729],
       [ 0.4911867 ,  0.79514022, -0.8467326 , ..., -0.52626175,
        -2.239829  ,  3.55134048],
       ...,
       [-1.26947162,  0.75921041,  0.85022184, ..., -1.03556061,
         0.0800678 , -0.27959729],
       [-1.16804239,  1.30742971, -0.8467326 , ...,  0.21453659,
         0.19107362, -0.27959729],
       [-0.19393904,  0.3083493 ,  0.37676218, ..., -1.39438481,
        -0.15067462, -0.27959729]])

## Split the data into training and testing data

In [11]:
import sklearn

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, target, test_size=0.3, random_state=1234)

## Build neural network

In [15]:
import keras

In [16]:
from keras.models import Sequential
from keras.layers import Dense

In [30]:
def regression_model():

    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

### Get input shape for the input layer

In [31]:
n_cols = X_train.shape[1]

In [32]:
model = regression_model()

## Train the network

In [33]:
model.fit(X_train, y_train, epochs=50)

Epoch 1/50
23/23 [==============================] - 0s 2ms/step - loss: 1479.3680
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1460.5347
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1441.6505
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1421.9216
Epoch 5/50
23/23 [==============================] - 0s 3ms/step - loss: 1401.5836
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1380.1180
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1358.1051
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1334.6230
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1310.0670
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1284.0283
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1256.3821
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 1227.4595
Epoch 13/50
2

### Evaluate using sklearn mean squared error

In [34]:
from sklearn.metrics import mean_squared_error

In [35]:
y_pred = model.predict(X_test)

In [36]:
mean_squared_error(y_test, y_pred)

209.26728449055136

## Repeat 50 times

In [37]:
def mse_list():
        
    #initialise the model
    model = regression_model()
    
    for i in range(50):
        #step 1
        X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=1234)
        #step 2
        model.fit(X_train, y_train, epochs=50)
        # step 3
        y_pred = model.predict(X_test)
        MSE = mean_squared_error(y_test, y_pred)
        MSE_list.append(MSE)

In [38]:
MSE_list = []

In [39]:
mse_list()

Epoch 1/50
23/23 [==============================] - 0s 2ms/step - loss: 7500.2192
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 3155.9309
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 2419.5876
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1934.3652
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1550.5348
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1232.1462
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 972.3453
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 774.4943
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 623.7463
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 518.5723
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 435.4518
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 377.4701
Epoch 13/50
23/23 [

In [40]:
MSE_list

[104.78368231844333,
 88.26933614739318,
 73.41501794391115,
 62.60060963106888,
 63.08488967843198,
 60.64744742299717,
 69.08142229159321,
 53.716355396495395,
 54.53883432480646,
 52.077405448217746,
 51.804984303834416,
 53.50337963980068,
 52.434142124985385,
 56.3977078389997,
 51.14415845835674,
 51.69655249777069,
 51.55397439989792,
 51.26755280842964,
 54.62457493143067,
 53.29491533437326,
 51.717371835603466,
 52.922968101499976,
 51.81978778767279,
 52.04532219135384,
 51.39058747339701,
 53.49933518822189,
 61.27336633588753,
 52.813770688925054,
 52.13051805445574,
 55.80286145243065,
 51.95152347492887,
 51.803365730121186,
 52.67526539532246,
 52.14644635532026,
 51.27648200836937,
 53.921560949297636,
 51.42037619593177,
 52.1090545069171,
 50.943700861867754,
 53.51983246227012,
 51.59363349474807,
 50.17737053638588,
 50.86569079398928,
 50.79458004901071,
 51.094406523609564,
 50.044199010567525,
 50.62018988141477,
 50.59305001076571,
 50.92346429854775,
 52.27477

## Finding mean

In [41]:
np.mean(MSE_list)

55.522036066761615

## Finding standard deviation

In [42]:
np.std(MSE_list)

9.708056362929982

## Concluson

The mean of mean squared errors in A (58.561) is higher than the mean of mean squared errors in B (55.522)